### ISSUE TITLE
Deferred foreign key checks in migrations can take a lot of time

### ISSUE THREAD

**Issue by juan-marin** [groue/GRDB.swift#1087]
### What did you do?

Hi, we are experiencing a very ugly crash with our app, any help will be greatly appreciated.

It's the second version that implements the GRDB framework to work with our SQLite DB. If the db is not full of data meaning less of 2 or 3 GB the migration runs smoothly, it's when there is a lot of data that the app crashes:

exhausted real (wall clock) time allowance of 19.05 seconds

The migration takes around 5 minutes with 13GBs of data so if the process is triggered from the didFinishLaunchingWithOptions the watchdog terminates the app. I have tried to do this migration using the asyncMigrate and wait until the process finishes and the app doesn't crash, but almost 5 minutes to complete this seems pretty excessive and the worst user experience. Am I doing something wrong? Is there anything I am missing?

var migrator = DatabaseMigrator()
        migrator.registerMigration( "accDb_v\( toVersion )" ) { ( db: Database ) in
            try? db.create(
                        table: "ProductContainers", body: { table in
                            table.column( "ContainerId", .text ).primaryKey().notNull().indexed()
                            table.column( "CEIProductId", .text )
                            table.column( "SerialNo", .text )
                            table.column( "GTIN", .text )
                            table.column( "DOM", .double )
                            table.column( "BatchNo", .text )
                            table.column( "ExpirationDate", .double )
                        }
                    )

              try? db.execute(sql: "UPDATE _version SET version = ?",  arguments: [toVersion])
        }

try? migrator.migrate( dbQueue, upTo: "v\( toVersion )" )

### What did you expect to happen?

Migration

### What happened instead?

Crash

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** 5.12.0
**Installation method:** SPM
**Xcode version:** 12.5
**Swift version:** 5
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** Big Sur 11.6

**Comment by juan-marin** [comment]
Hi @groue, I just wanted to give an update to my issue, I have found and fixed the problem so if some one else runs with the same problem they know where to look.

For large DB with large datasets the DB connection needs to be done with the foreign keys disabled if not GRDB runs a "PRAGMA foreign_key_check" and that process is what it was taking up to 5mins for my 13GB database.

Hope it helps anyone else.

**Comment by groue** [comment]
Hello @juan-marin, thanks for the feedback!

> The migrator needs to be done with the foreign keys disabled if not GRDB runs a "PRAGMA foreign_key_check" and that process is what it was taking up to 5mins for my 13GB database.

Oh, now I understand. The detailed explanation is:

- SQLite documents, in [Making Other Kinds Of Table Schema Changes](https://www.sqlite.org/lang_altertable.html#making_other_kinds_of_table_schema_changes), that some schema alterations require to rebuild whole tables. Such alterations must run with foreign key checks disabled. It is recommended to run `PRAGMA foreign_key_check` before committing.
- GRDB migrations use this technique, when the database has foreign keys enabled. This allows migrations to support all kinds of schema change, including those that require tables to be rebuilt. This is handy.

When you disable foreign keys before running migrations, GRDB does not have to disable foreign keys and check them later: this is faster.

_But this is also quite dangerous_: each migration will commit its change on disk without checking foreign keys. You may end up with a corrupt database.

A better option could be to instruct GRDB to _not_ disable and check foreign keys for your migration. This is not currently possible, but this is a feature that could be discussed, and added, if it makes a noticeable difference in performance.

**Comment by juan-marin** [comment]
Yes, that would be great! 

Thank you

**Comment by groue** [comment]
I mean telling a migration that it should run without disabling foreign keys and without checking them before committing. It would thus run with foreign keys enabled if the migrated database has foreign keys enabled.

The flag that disables foreign key checks, you found it already: it is running migrations on a database with disabled foreign keys. This is a dangerous practice that can easily corrupt data, so it won't be fostered.

**Comment by juan-marin** [comment]
OK yeah I totally understand you now, I really think this is needed. More for users with really big Databases meaning lots of tables, the one that I am working on has around 150  and there are users that can grow their data up to 15GB. And I totally agree, doing the migration with out foreign keys is very dangerous but right now we do not have another option, or at least I have not found it, is there anything different you think we could be doing?

**Comment by groue** [comment]
We could imagine the following api:

```swift
// Current and default behavior.
// Can perform all schema changes.
migrator.registerMigration("v1", foreignKeyChecks: .deferred) { db in ... }

// Don't disable foreign keys.
// Some schema changes are not possible.
migrator.registerMigration("v1", foreignKeyChecks: .immediate) { db in ... }

// DANGEROUS: Disable foreign keys and do not check them.
// Can perform all schema changes. Can corrupt the database.
migrator.registerMigration("v1", foreignKeyChecks: .disabled) { db in ... }
```

All those flags are ignored if migrations run of a database with disabled foreign keys.

**Comment by juan-marin** [comment]
That sounds like a good plan to me. 
 
Thank you!

**Comment by groue** [comment]
All right, @juan-marin, thanks for your input :-)

**Comment by groue** [comment]
I'll look at this precisely in the near future.

**Comment by juan-marin** [comment]
That is awesome, please let me know if there is anything I can do to help.

**Comment by groue** [comment]
Yes you do, @juan-marin :-)

- Please use the [dev/foreign-keys-checks](https://github.com/groue/GRDB.swift/tree/dev/foreign-keys-checks) branch
- Read its [updated documentation](https://github.com/groue/GRDB.swift/blob/dev/foreign-keys-checks/Documentation/Migrations.md#foreign-key-checks).
- Please test mitigation technique 1 before testing mitigation technique 2. Please report performance improvements brought by technique 1 before reporting performance improvements brought by technique 2.

Thanks a lot for your help!

**Comment by juan-marin** [comment]
Hi @groue,

So I realized the tests with the 13GB database I've been working on, results as follow:

Technique 1 -> .immediate Migration Took ---> 0.0019530057907104492 seconds
                         .deferred Migration Took ---> 250.39679396152496 seconds
Technique 2 -> Migration Took ---> 0.004932999610900879 seconds

I don't see why someone will take technique 2 as their path. Anyways I think this fixes all the issues encountered.

When will you push this changes to master?

Thanks you

**Comment by groue** [comment]
That's great, @juan-marin, thank you very much for running those performance tests 👍

Using `.immediate` foreign key checks vastly improves the speed of your migration AND makes sure database integrity is preserved. That's the best outcome we could hope for 🎉 

The second technique, disabling deferred key checks, will be needed by applications that run migrations incompatible with immediate foreign key checks. I'm happy you do not need it, because disabling foreign key checks is dangerous!

> When will you push this changes to master?

I generally provide no ETA, but you shouldn't have to wait too long. Meanwhile you can use the branch.

**Comment by juan-marin** [comment]
Thank you!

**Comment by groue** [comment]
The PR is almost ready: https://github.com/groue/GRDB.swift/pull/1095

**Comment by groue** [comment]
The feature has shipped in v5.13.0 - [documentation](https://github.com/groue/GRDB.swift/blob/v5.13.0/Documentation/Migrations.md#foreign-key-checks)

**Comment by juan-marin** [comment]
Thank you!
